### PR TITLE
feat(decide): Adding support for decide options and decision reasons.

### PR DESCRIPTION
**Pull request by yasirfolio3**
## Summary
A part of multiple PRs for Decide API:
- added DecideOptions.
- added DecisionReasons.
- Refactored interfaces with additional parameter of reasons and options.
- Fixed unit tests throughout the sdk. 

## Tests
- Unit tests added.

### CODE REVIEW

File: pkg/decision/experiment_bucketer_service.go [hunk 1]
```diff
@@ -45,19 +46,19 @@ func NewExperimentBucketerService(logger logging.OptimizelyLogProducer) *Experim
...
 	if experiment.AudienceConditionTree != nil {
 		condTreeParams := entities.NewTreeParameters(&userContext, decisionContext.ProjectConfig.GetAudienceMap())
 		s.logger.Debug(fmt.Sprintf(logging.EvaluatingAudiencesForExperiment.String(), experiment.Key))
-		evalResult, _ := s.audienceTreeEvaluator.Evaluate(experiment.AudienceConditionTree, condTreeParams)
+		evalResult, _ := s.audienceTreeEvaluator.Evaluate(experiment.AudienceConditionTree, condTreeParams, reasons)
 		s.logger.Debug(fmt.Sprintf(logging.ExperimentAudiencesEvaluatedTo.String(), experiment.Key, evalResult))
 		if !evalResult {
 			s.logger.Debug(fmt.Sprintf(logging.UserNotInExperiment.String(), userContext.ID, experiment.Key))
-			experimentDecision.Reason = reasons.FailedAudienceTargeting
+			experimentDecision.Reason = pkgReasons.FailedAudienceTargeting
```
Comment (jaeopt) [inline pkg/decision/experiment_bucketer_service.go:61]: This message should go into new DecisionReasons as well.
Wondering how we can merge this existing "Reason" with new DecisionReasons :)

File: pkg/decision/persisting_experiment_service_test.go [hunk 2]
```diff
@@ -53,12 +56,14 @@ func (s *PersistingExperimentServiceTestSuite) SetupTest() {
 	s.testComputedDecision = ExperimentDecision{
 		Variation: &computedVariation,
 	}
-	s.mockExperimentService.On("GetDecision", s.testDecisionContext, testUserContext).Return(s.testComputedDecision, nil)
+	s.options = decide.OptimizelyDecideOptions{}
+	s.reasons = decide.NewDecisionReasons(s.options)
+	s.mockExperimentService.On("GetDecision", s.testDecisionContext, testUserContext, s.options, s.reasons).Return(s.testComputedDecision, nil)
 }
 
 func (s *PersistingExperimentServiceTestSuite) TestNilUserProfileService() {
```
Comment (jaeopt) [inline pkg/decision/persisting_experiment_service_test.go:64]: Not sure if it's planned to add later, but we need test cases to validate "IgnoreUPS" option here.

**Review by jaeopt — CHANGES_REQUESTED** [review]
I understand that this PR extends existing interface to support changes for new decide APIs. It looks good in that sense. 
Not sure if you plan to add them later, but I do not see collecting log messages into DecisionReasons.

File: pkg/decision/persisting_experiment_service_test.go (hunk 2 above)
Comment (yasirfolio3) [reply pkg/decision/persisting_experiment_service_test.go:64]: yes these tests are included in the succeeding PR. #299

File: pkg/decide/default_decision_reasons.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	}
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+	o.logs = append(o.logs, message)
+	return message
+}
+
+// ToReport returns reasons to be reported.
+func (o *DefaultDecisionReasons) ToReport() []string {
+	o.mutex.RLock()
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply pkg/decide/default_decision_reasons.go:63]: this was added as a precautionary measure for scenario's where user might call `ToReport` while a go-routine is adding `errors` or `logs`.

File: pkg/decision/composite_feature_service.go [hunk 4]
```diff
@@ -20,20 +20,21 @@ package decision
...
+	"github.com/optimizely/go-sdk/pkg/decide"
 	"github.com/optimizely/go-sdk/pkg/entities"
 	"github.com/optimizely/go-sdk/pkg/logging"
 )
 
 // CompositeFeatureService is the default out-of-the-box feature decision service
 type CompositeFeatureService struct {
 	featureServices []FeatureService
-	logger logging.OptimizelyLogProducer
+	logger          logging.OptimizelyLogProducer
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply pkg/decision/composite_feature_service.go:31]: this indentation fix is on purpose, it should be aligned with the text above.

File: pkg/decision/evaluator/condition.go [hunk 5]
```diff
@@ -61,7 +61,7 @@ func (c CustomAttributeConditionEvaluator) Evaluate(condition entities.Condition
 		return false, fmt.Errorf(`invalid Condition matcher "%s"`, condition.Match)
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply pkg/decision/evaluator/condition.go:61]: @msohailhussain

File: pkg/decision/evaluator/condition.go [hunk 6]
```diff
@@ -41,11 +42,10 @@ func NewCustomAttributeConditionEvaluator(logger logging.OptimizelyLogProducer)
...
 
 // Evaluate returns true if the given user's attributes match the condition
-func (c CustomAttributeConditionEvaluator) Evaluate(condition entities.Condition, condTreeParams *entities.TreeParameters) (bool, error) {
+func (c CustomAttributeConditionEvaluator) Evaluate(condition entities.Condition, condTreeParams *entities.TreeParameters, reasons decide.DecisionReasons) (bool, error) {
 	// We should only be evaluating custom attributes
 
 	if condition.Type != customAttributeType {
-
 		c.logger.Warning(fmt.Sprintf(logging.UnknownConditionType.String(), condition.StringRepresentation))
 		return false, fmt.Errorf(`unable to evaluate condition of type "%s"`, condition.Type)
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply pkg/decision/evaluator/condition.go:50]: @msohailhussain

**Comment by yasirfolio3** [comment]
> I understand that this PR extends existing interface to support changes for new decide APIs. It looks good in that sense.
> Not sure if you plan to add them later, but I do not see collecting log messages into DecisionReasons.

yes i was waiting for one of the PR's on other sdk's to finalize so that i can add messages to decision reasons, will do that in another PR.

File: pkg/decision/experiment_bucketer_service.go (hunk 1 above)
Comment (yasirfolio3) [inline pkg/decision/experiment_bucketer_service.go:61]: This comment was originally added by @jaeopt but accidentally deleted:

This message should go into new DecisionReasons as well.
Wondering how we can merge this existing "Reason" with new DecisionReasons.

Comment (yasirfolio3) [reply pkg/decision/experiment_bucketer_service.go:61]: @msohailhussain

**Review by jaeopt — APPROVED** [review]
LGTM - I expect all reasons collecting in new PRs

**Review by pawels-optimizely — CHANGES_REQUESTED** [review]
reasons are passed to Evaluator, condition and Matchers, but I don't think they are used. 
Please either clean them up  (remove them) or use them.

**Comment by yasirfolio3** [comment]
> reasons are passed to Evaluator, condition and Matchers, but I don't think they are used.
> Please either clean them up (remove them) or use them.

We will be using them in a separate PR just dedicated to reasons with all the unit tests. 👍

File: pkg/decide/decision_reasons_test.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+func TestInfoLogsAreOnlyReportedWhenIncludeReasonsOptionIsSet(t *testing.T) {
+	options := OptimizelyDecideOptions{
+		DisableDecisionEvent:     true,
+		EnabledFlagsOnly:         true,
+		IgnoreUserProfileService: true,
+		ExcludeVariables:         true,
+	}
+	reasons := NewDecisionReasons(options)
+	reasons.AddError("error message")
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply pkg/decide/decision_reasons_test.go:54]: done 👍

File: pkg/decide/default_decision_reasons.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pawels-optimizely) [reply pkg/decide/default_decision_reasons.go]: why do we need mutex here ?

File: pkg/decision/evaluator/condition.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (pawels-optimizely) [reply pkg/decision/evaluator/condition.go:61]: we can probably put it within the reasons

File: pkg/decision/experiment_bucketer_service.go (hunk 1 above)
Comment (pawels-optimizely) [reply pkg/decision/experiment_bucketer_service.go:61]: yeah, these make sense

**Review by pawels-optimizely — APPROVED** [review]

**Review by msohailhussain — APPROVED** [review]
lgtm

**Pull request merged**